src/main/kotlin/org/team4099/lib/pathfollow/AStar.kt

```diff
@@ -0,0 +1,149 @@
+          .minByOrNull { (_, pose) -> (startingPosition.x - pose.x).absoluteValue.inMeters }
+          ?.index
+          ?: 0
+      val startingNode = Node(null, grid[bestYStarting][bestXStarting])
+
+      val bestYEnding =
+        yPoints
+          .withIndex()
+          .minByOrNull { (_, y) -> (endingPosition.y - y).absoluteValue.inMeters }
+          ?.index
+          ?: 0
+      val bestXEnding =
+        grid[bestYEnding]
+          .withIndex()
+          .minByOrNull { (_, pose) -> (endingPosition.x - pose.x).absoluteValue.inMeters }
+          ?.index
+          ?: 0
+      val endingNode = Node(null, grid[bestYEnding][bestXEnding])
+
+      // Other variables instantiated for later use when running the algorithm.
+      val deltas =
+        listOf(
+          listOf(0, -1),
+          listOf(0, 1),
+          listOf(-1, 0),
+          listOf(1, 0),
+          listOf(-1, -1),
+          listOf(1, 1),
+          listOf(-1, 1),
+          listOf(1, -1)
+        )
+      val openList = mutableListOf<Node>(startingNode)
+      val closedList = mutableListOf<Node>(endingNode)
+      var currentNode: Node?
+
+      while (openList.size >=
+        1
+      ) { // Keep running until all the nodes are visited. In that case, no optimal path is
+        // found.
+        currentNode =
+          openList.minByOrNull { it.f } ?: openList.last() // Find the node with the lowest cost.
+
+        openList.remove(currentNode)
+        closedList.add(currentNode)
+
+        // Path found (current node with the lowest cost is the ending node).
+        if (currentNode == endingNode) {
+          val path = mutableListOf<Pose2d>()
+
+          while (currentNode != null) {
+            path.add(currentNode.position)
+            currentNode = currentNode.parent
+          }
+
+          return path.reversed()
+        }
+
+        val currentY = yPoints.indexOfFirst { it.epsilonEquals(currentNode.position.y) }
+        val currentX = grid[currentY].indexOfFirst { it.x.epsilonEquals(currentNode.position.x) }
+
+        // Iterate through all the adjacent nodes.
+        for ((deltaX, deltaY) in deltas) {
+          val newY = currentY + deltaY
+          val newX = currentX + deltaX
+
+          // Suggested adjacent node is out of range.
+          if (!(0 <= newY && newY < grid.size) || !(0 <= newX && newX < grid[newY].size)) {
+            continue
+          }
+
+          val newNode = Node(currentNode, grid[newY][newX])
+          newNode.g = currentNode.g + 1
+          newNode.h =
+            (
+              (currentNode.position.x - newNode.position.x).absoluteValue.inMeters.pow(2.0) +
+                (currentNode.position.y - newNode.position.y).absoluteValue.inMeters.pow(2.0)
+              )
+              .toInt()
+          newNode.f = newNode.g + newNode.h
+
+          // Discard node if it means going over the charge station.
+          if ((newNode.position.x - currentNode.position.x).absoluteValue.inMeters >= 10) {
```

Comment:
old game so understandable but i think you want to construct your graph so that this isn't even a possibility. think about an actual graph, it can only follow edges that exist, an edge that goes over the charge station shouldn't exist